[WinSheng1] iP

WinSheng1 opened the pull request:
# TaskBot
> Don't worry, **be happy**.

TaskBot is your personal task management companion, but Microsoft To Do is better... so don't use it, but if you really want to, here are it's features!
### Features
- Create tasks
- Delete tasks

To take up more storage space in your computer, all you need to do is,

1. Download it from [here](https://todo.microsoft.com/tasks/)
2. Thank me for not wasting your storage space
3. ~~_Paylah paynow me_~~ :see_no_evil:

- [x] Task
- [x] List

Some code:
```
public static void main(String[] args) {
    new Duke("./data/duke.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

WinSheng1 committed:
Level 0. Rename, Greet, Exit

WinSheng1 committed:
Level 1. Echo

WinSheng1 committed:
Level 2. Add, List

WinSheng1 committed:
Level 3. Mark as Done

WinSheng1 committed:
Level 4. ToDos, Events, Deadlines

WinSheng1 committed:
Level 4. ToDos, Events, Deadlines

WinSheng1 committed:
A-TextUiTesting

WinSheng1 committed:
Merge remote-tracking branch 'origin/master'

WinSheng1 committed:
Level 5. Handle Errors

WinSheng1 committed:
Level 6. Delete

WinSheng1 committed:
Updated parts of code for readability

WinSheng1 committed:
Level 7. Save

WinSheng1 committed:
Level 7. Save

WinSheng1 committed:
Merge branch 'branch-Level-7'

WinSheng1 committed:
Level 7. Save

WinSheng1 committed:
Level 8. Dates and Times

WinSheng1 committed:
Merge branch 'branch-Level-8'

WinSheng1 committed:
A-MoreOOP

WinSheng1 committed:
Merge branch 'branch-A-MoreOOP'

WinSheng1 committed:
A-Packages

WinSheng1 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

WinSheng1 committed:
A-Gradle

WinSheng1 committed:
A-JUnit

WinSheng1 committed:
A-JavaDoc

WinSheng1 committed:
A-CodingStandard

WinSheng1 committed:
Level 9. Find

WinSheng1 committed:
Merge branch 'branch-A-JavaDoc'

WinSheng1 committed:
resolving conflicts for A-CodingStandard

WinSheng1 committed:
resolving conflicts for Level-9

WinSheng1 committed:
refactored files, modified and added JavaDocs

sk2001git commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+ * The Duke class represents the TaskMate bot that allows users to
+ * manage tasks by adding, deleting, marking and un-marking tasks.
+ *
+ * @author Win Sheng
+ * @since 3 September 2023
+ */
```
I like how you updated the java docs to include date updated 😃 
This was done for all documents which is consistent!

sk2001git commented on src/main/java/duke/exception/InvalidCommandException.java:
```diff
@@ -0,0 +1,25 @@
+     * Constructs a new InvalidCommandException with a default error message and the list of available commands.
+     */
+    public InvalidCommandException() {
+        super("Invalid Command.\n"
+                + "Below are the list of available commands:\n"
+                + "list - displays the list of tasks\n"
+                + "mark x - marks task x as done\n"
+                + "unmark x - marks task x as undone\n"
+                + "delete x - deletes task x\n"
+                + "todo x - creates task with description x\n"
+                + "deadline x /by y - creates task with description x by deadline y\n"
+                + "event x /from y /to z - creates task with description x from y to z");
```
I like how the + adheres to the coding standard which is always on the next line 😄

sk2001git commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+package duke.task;
```
I think every class has a package, which adheres to the guideline 👍

sk2001git commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    @Override
+    public String toFileString() {
+        return type() + " | " + (isDone ? "1" : "0") + " | " + description + " | " + fromDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm")) + " | " + toDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
```
I think the indentation for the line might be a bit off, have noticed this in the line below as well!

sk2001git commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public abstract class Task {
+    public String description;
+    public boolean isDone;
```
I think all your variables involving booleans are named correctly with the is prefix, well done! 👍

sk2001git commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,319 @@
+                String lowerCapsFirstWord = firstWord.toLowerCase();
+
+                try {
+                    switch (lowerCapsFirstWord) {
+                        case "bye":
+                            ui.showGoodbye();
+                            Duke.isDone = true;
+                            break;
+                        case "list":
+                            listTasks(taskList, ui);
+                            break;
+                        case "mark":
+                            markTask(userInput, taskList, ui);
+                            break;
+                        case "unmark":
+                            unmarkTask(userInput, taskList, ui);
+                            break;
+                        case "delete":
+                            deleteTask(userInput, taskList, ui);
+                            break;
+                        case "todo":
+                            addTodoTask(userInput, taskList, ui);
+                            break;
+                        case "deadline":
+                            addDeadlineTask(userInput, taskList, ui);
+                            break;
+                        case "event":
+                            addEventTask(userInput, taskList, ui);
+                            break;
+                        case "find":
+                            findByKeyword(userInput, taskList, ui);
```
I think according to the coding standard, case statement belongs on the same line with switch statement?
I think this is what it should look like!
```suggestion
                    switch (lowerCapsFirstWord) {
                    case "bye":
                        ui.showGoodbye();
                        Duke.isDone = true;
                        break;
                    case "list":
                        listTasks(taskList, ui);
                        break;
                    case "mark":
                        markTask(userInput, taskList, ui);
                        break;
                    case "unmark":
                        unmarkTask(userInput, taskList, ui);
                        break;
                    case "delete":
                        deleteTask(userInput, taskList, ui);
                        break;
                    case "todo":
                        addTodoTask(userInput, taskList, ui);
                        break;
                    case "deadline":
                        addDeadlineTask(userInput, taskList, ui);
                        break;
                    case "event":
                        addEventTask(userInput, taskList, ui);
                        break;
                    case "find":
                        findByKeyword(userInput, taskList, ui);
```

sk2001git commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,102 @@
+package duke.util;
+
+import duke.task.Task;
+import duke.task.ToDo;
+import duke.task.Deadline;
+import duke.task.Event;
+
+import java.io.IOException;
+import java.io.File;
+import java.io.FileReader;
+import java.io.BufferedReader;
+import java.io.FileWriter;
+import java.util.ArrayList;
```
I like how the imports are done individually as specified in the coding standard!

sk2001git replied:
I think that the importing statements may or may not be inconsistent,  i think you can use the checkstyle to confirm it!

sk2001git commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,102 @@
+import java.io.BufferedReader;
+import java.io.FileWriter;
+import java.util.ArrayList;
+
+/**
+ * The Storage class handles loading and saving of tasks to a file specified by a file path.
+ *
+ * @author Win Sheng
+ * @since 3 September 2023
+ */
```
Overall, all classes have documentation along with constructors and methods! Well done 👍

sk2001git reviewed (commented):
LGTM. There are some nits for coding standards like switch statements however in general nothing major!

JasonRay168 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+     * @param description The description of the task.
+     * @param by The deadline of the task.
+     */
+    public Deadline(String description, String by) {
```
Maybe change "by" to something else, like "deadline" to make it clearer.

JasonRay168 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,76 @@
+     * @param from The start date and time of the task.
+     * @param to The end date and time of the task.
+     */
+    public Event(String description, String from, String to) {
```
Personally, I think it's better to name the from and to into start and end, since they're both nouns (same argument as the  by)

JasonRay168 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    public boolean containsKeyword(String keyword) {
+        return description.toLowerCase().contains(keyword.toLowerCase());
+    }
```
I like how you also remembered to use words that convey yes or no questions for the boolean methods as well

JasonRay168 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,319 @@
+            }
+        } catch (InvalidCommandException e) {
+            ui.showMessage(e.getMessage());
+        }
```
This entire try catch block also has another try catch block inside. Combined with if else, it appears deeply nested. Maybe can refactor a bit to make it one try catch block outside the if else instead of nesting another try catch inside.

JasonRay168 commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,102 @@
+import duke.task.Deadline;
+import duke.task.Event;
+
+import java.io.IOException;
+import java.io.File;
+import java.io.FileReader;
+import java.io.BufferedReader;
+import java.io.FileWriter;
+import java.util.ArrayList;
```
To also add on to @sk2001git, it would be better to arrange the import statements in alphabetical order. This adheres to the checkstyle. So lines 8 to 13 would be:
```ruby
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

import java.util.ArrayList;
```

JasonRay168 reviewed (commented):
Overall, you did a good job to maintain your coding quality. Outside a few small issues, it's very consistent and looks good.

WinSheng1 committed:
Level 10. GUI

WinSheng1 committed:
Merge branch 'branch-Level-10'

WinSheng1 committed:
reverting back

WinSheng1 committed:
A-Assertions: Use Assertions

WinSheng1 committed:
Merge pull request #1 from WinSheng1/branch-A-Assertions

A-Assertions: Use Assertions

WinSheng1 committed:
Improved code based on code quality

There are several methods that were using duplicated code.

Unifying these duplicated code lines allow for better readability.

This was resolved by creating a method and replacing those duplicated code, making the code cleaner and easier to read.

Some lines were also realigned for better flow and readability.

WinSheng1 committed:
Merge pull request #2 from WinSheng1/branch-A-CodeQuality

Improved code based on code quality

WinSheng1 committed:
Reminders

Reminders are used to remind users of tasks due within 3 days.

Reminders will be automatically printed on start up of application.

WinSheng1 committed:
Merge branch 'branch-B-Reminders'

WinSheng1 committed:
Updated README as User Guide

WinSheng1 committed:
Minor changes to file creation

WinSheng1 committed:
Updated javadocs
